btcutil/psbt/taproot.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// TaprootTapLeafScriptSorter implements sort.Interface for the
+// TaprootTapLeafScript struct.
+type TaprootTapLeafScriptSorter []*TaprootTapLeafScript
+
+func (s TaprootTapLeafScriptSorter) Len() int { return len(s) }
+
+func (s TaprootTapLeafScriptSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s TaprootTapLeafScriptSorter) Less(i, j int) bool {
+	return bytes.Compare(s[i].ControlBlock, s[j].ControlBlock) < 0
+}
+
+// TaprootBip32Derivation encapsulates the data for the input and output
+// taproot specific BIP-32 derivation key-value fields.
+type TaprootBip32Derivation struct {
+	// XOnlyPubKey is the raw public key serialized in the x-only BIP-340
+	// format.
+	XOnlyPubKey []byte
+
+	// LeafHashes is a list of leaf hashes that the given public key is
+	// involved in.
+	LeafHashes [][]byte
+
+	// MasterKeyFingerprint is the fingerprint of the master pubkey.
+	MasterKeyFingerprint uint32
+
+	// Bip32Path is the BIP 32 path with child index as a distinct integer.
+	Bip32Path []uint32
+}
+
+// TaprootBip32DerivationSorter implements sort.Interface for the
+// TaprootBip32Derivation struct.
+type TaprootBip32DerivationSorter []*TaprootBip32Derivation
+
+func (s TaprootBip32DerivationSorter) Len() int { return len(s) }
+
+func (s TaprootBip32DerivationSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s TaprootBip32DerivationSorter) Less(i, j int) bool {
+	return bytes.Compare(s[i].XOnlyPubKey, s[j].XOnlyPubKey) < 0
+}
+
+// readTaprootBip32Derivation deserializes a byte slice containing the Taproot
+// BIP32 derivation info that consists of a list of leaf hashes as well as the
+// normal BIP32 derivation info.
+func readTaprootBip32Derivation(xOnlyPubKey,
+	value []byte) (*TaprootBip32Derivation, error) {
+
+	// The taproot key BIP 32 derivation path is defined as:
+	//   <hashes len> <leaf hash>* <4 byte fingerprint> <32-bit uint>*
+	// So we get at least 5 bytes for the length and the 4 byte fingerprint.
+	if len(value) < 5 {
+		return nil, ErrInvalidPsbtFormat
+	}
+
+	// The first element is the number of hashes that will follow.
+	reader := bytes.NewReader(value)
+	numHashes, err := wire.ReadVarInt(reader, 0)
+	if err != nil {
+		return nil, ErrInvalidPsbtFormat
+	}
+
+	// A hash is 32 bytes in size, so we need at least numHashes*32 + 5
+	// bytes to be present.
+	if len(value) < (int(numHashes)*32)+5 {
+		return nil, ErrInvalidPsbtFormat
+	}
+
+	derivation := TaprootBip32Derivation{
+		XOnlyPubKey: xOnlyPubKey,
+		LeafHashes:  make([][]byte, int(numHashes)),
+	}
+
+	for i := 0; i < int(numHashes); i++ {
+		derivation.LeafHashes[i] = make([]byte, 32)
+		n, err := reader.Read(derivation.LeafHashes[i])
+		if err != nil || n != 32 {
+			return nil, ErrInvalidPsbtFormat
+		}
+	}
+
+	// Extract the remaining bytes from the reader (we don't actually know
+	// how many bytes we read due to the compact size integer at the
+	// beginning).
+	leftover := make([]byte, reader.Len())
+	_, err = reader.Read(leftover)
```

Comment:
I think we can do this in place: 
```go
io.Copy(io.Discard, reader)
```

Reply:
I didn't know about `io.Discard`. But in this case we actually want to get the remaining bytes, we just don't know how many there are. I went with a `WriteTo` approach instead, hoping this makes it a bit more clear.